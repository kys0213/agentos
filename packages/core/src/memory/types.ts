export type NodeType = 'query'|'answer'|'feedback'|'entity';
export type EdgeType = 'similar_to'|'responded_with'|'has_feedback'|'refers_to_entity';

export interface BaseNode {
  id: string;
  type: NodeType;
  text?: string;
  canonicalKey?: string;
  embedding?: number[] | Map<number, number>;
  createdAt: number;
  lastAccess: number;
  weights: { repeat: number; feedback: number };
  degree: number;
  pinned?: boolean;
}

export interface Edge {
  id: string;
  from: string;
  to: string;
  type: EdgeType;
  weight: number;
  createdAt: number;
  lastAccess: number;
}

export interface GraphConfig {
  maxNodes: number;
  maxEdges: number;
  halfLifeMin: number;
  tauDup: number;
  tauSim: number;
  protectMinDegree: number;
  enableInvertedIndex?: boolean;
}

export interface EmbeddingProvider {
  embed(text: string): number[] | Map<number, number>;
  exportState?(): any;
  importState?(s: any): void;
}

