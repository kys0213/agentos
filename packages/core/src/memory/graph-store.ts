import fs from 'fs/promises';
import { BaseNode, Edge, EdgeType, EmbeddingProvider, GraphConfig } from './types';
import { SimpleEmbedding, SparseVec, normalize } from './embedding/simple-embedding';
import { cosineSparse } from './utils/cosine-sparse';
import { canonicalKey, defaultCanonicalMeta } from './utils/canonical-key';
import { serializeSparse, deserializeSparse } from './utils/serialize-sparse';

export class GraphStore {
  private nodes = new Map<string, BaseNode>();
  private edges = new Map<string, Edge>();
  private byCanonical = new Map<string, string>();
  private inverted?: Map<string, Set<string>>;
  private dirty = false;

  constructor(
    private cfg: GraphConfig,
    private embedder: EmbeddingProvider = new SimpleEmbedding()
  ) {
    if (cfg.enableInvertedIndex) {
      this.inverted = new Map();
    }
  }

  private indexNode(id: string, text?: string, canonical?: string) {
    if (canonical) {
      this.byCanonical.set(canonical, id);
    }
    if (!this.inverted || !text) {
      return;
    }
    const grams = normalize(text).split(' ');
    for (const g of grams) {
      if (!g) {
        continue;
      }
      const set = this.inverted.get(g) ?? new Set();
      set.add(id);
      this.inverted.set(g, set);
    }
  }

  upsertQuery(text: string) {
    const now = Date.now();
    const emb = this.embedder.embed(text) as SparseVec;
    const ck = canonicalKey(text);
    const existing = this.byCanonical.get(ck);
    if (existing) {
      const q = this.nodes.get(existing)!;
      q.weights.repeat += 1;
      q.lastAccess = now;
      this.touchNeighbors(existing, now);
      this.evictIfNeeded();
      return q.id;
    }
    const cand: { id: string; sim: number }[] = [];
    for (const n of this.nodes.values()) {
      if (n.type !== 'query' || !n.embedding) {
        continue;
      }
      const sim = cosineSparse(emb, n.embedding as SparseVec);
      cand.push({ id: n.id, sim });
    }
    cand.sort((a, b) => b.sim - a.sim);
    const best = cand[0];
    if (best && best.sim >= this.cfg.tauDup) {
      const q = this.nodes.get(best.id)!;
      q.weights.repeat += 1;
      q.lastAccess = now;
      this.touchNeighbors(q.id, now);
      this.evictIfNeeded();
      return q.id;
    }
    const id = `q_${now}_${Math.random().toString(36).slice(2, 8)}`;
    const q: BaseNode = {
      id,
      type: 'query',
      text,
      canonicalKey: ck,
      embedding: emb,
      createdAt: now,
      lastAccess: now,
      degree: 0,
      weights: { repeat: 0, feedback: 0 },
    };
    this.nodes.set(id, q);
    this.indexNode(id, text, ck);
    for (const c of cand.slice(0, 10)) {
      if (c.sim >= this.cfg.tauSim) {
        this.link(id, c.id, 'similar_to', c.sim);
        this.bump(id, +0.3);
        this.bump(c.id, +0.3);
      }
    }
    this.dirty = true;
    this.evictIfNeeded();
    return id;
  }

  recordFeedback(targetQueryId: string, label: 'up' | 'down' | 'retry', note?: string) {
    const now = Date.now();
    const id = `f_${now}_${Math.random().toString(36).slice(2, 6)}`;
    const f: BaseNode = {
      id,
      type: 'feedback',
      text: `feedback:${label}${note ? ' ' + note : ''}`,
      createdAt: now,
      lastAccess: now,
      degree: 0,
      weights: { repeat: 0, feedback: 0 },
    };
    this.nodes.set(id, f);
    this.link(targetQueryId, id, 'has_feedback', 1.0);
    if (label === 'up') {
      this.bump(targetQueryId, +1.0);
    }
    if (label === 'down') {
      this.bump(targetQueryId, -1.0);
    }
    if (label === 'retry') {
      this.bump(targetQueryId, -0.5);
    }
    this.dirty = true;
    this.evictIfNeeded();
    return id;
  }

  adjustWeights(nodeId: string, deltas: { feedbackDelta?: number; repeatDelta?: number }) {
    const n = this.nodes.get(nodeId);
    if (!n) {
      return;
    }
    if (deltas.feedbackDelta) {
      n.weights.feedback += deltas.feedbackDelta;
    }
    if (deltas.repeatDelta) {
      n.weights.repeat += deltas.repeatDelta;
    }
    n.lastAccess = Date.now();
    this.dirty = true;
  }

  link(from: string, to: string, type: EdgeType, weight: number) {
    const now = Date.now();
    if (!this.nodes.has(from) || !this.nodes.has(to)) {
      return;
    }
    const key = `${from}::${type}::${to}`;
    const prev = this.edges.get(key);
    const e: Edge = {
      id: key,
      from,
      to,
      type,
      weight: (prev?.weight ?? 0) + weight,
      createdAt: prev?.createdAt ?? now,
      lastAccess: now,
    };
    if (!prev) {
      const a = this.nodes.get(from)!;
      a.degree++;
      a.lastAccess = now;
      const b = this.nodes.get(to)!;
      b.degree++;
      b.lastAccess = now;
    }
    this.edges.set(key, e);
  }

  searchSimilarQueries(text: string, k = 5) {
    const emb = this.embedder.embed(text) as SparseVec;
    const out: { id: string; sim: number; score: number; text?: string; canonicalKey?: string }[] =
      [];
    const ids = new Set<string>();
    if (this.inverted) {
      const grams = normalize(text).split(' ');
      for (const g of grams) {
        for (const id of this.inverted.get(g) ?? []) {
          ids.add(id);
        }
      }
    } else {
      for (const id of this.nodes.keys()) {
        ids.add(id);
      }
    }
    for (const id of ids) {
      const n = this.nodes.get(id)!;
      if (n.type !== 'query' || !n.embedding) {
        continue;
      }
      const sim = cosineSparse(emb, n.embedding as SparseVec);
      const score = 0.7 * sim + 0.3 * this.rank(n);
      out.push({ id, sim, score, text: n.text, canonicalKey: n.canonicalKey });
    }
    return out.sort((a, b) => b.score - a.score).slice(0, k);
  }

  private bump(nodeId: string, delta: number) {
    const n = this.nodes.get(nodeId);
    if (!n) {
      return;
    }
    n.weights.feedback += delta;
    n.lastAccess = Date.now();
  }
  private touchNeighbors(nodeId: string, ts: number) {
    for (const e of this.edges.values()) {
      if (e.from === nodeId || e.to === nodeId) {
        e.lastAccess = ts;
      }
    }
  }
  private rank(n: BaseNode) {
    const now = Date.now();
    const ageMin = (now - n.lastAccess) / (1000 * 60);
    const recency = Math.exp(-Math.log(2) * (ageMin / this.cfg.halfLifeMin));
    const repeat = Math.tanh((n.weights.repeat ?? 0) / 5);
    const fb = Math.tanh((n.weights.feedback ?? 0) / 5);
    return 0.5 * recency + 0.3 * repeat + 0.2 * fb;
  }
  private evictIfNeeded() {
    const { maxNodes, maxEdges, protectMinDegree } = this.cfg;
    const overN = this.nodes.size - maxNodes;
    if (overN > 0) {
      const cand: BaseNode[] = [];
      for (const n of this.nodes.values()) {
        if (n.pinned) {
          continue;
        }
        if ((n.degree ?? 0) >= protectMinDegree) {
          continue;
        }
        cand.push(n);
      }
      cand.sort((a, b) => this.rank(a) - this.rank(b));
      for (let i = 0; i < overN && i < cand.length; i++) {
        this.removeNode(cand[i].id);
      }
    }
    const overE = this.edges.size - maxEdges;
    if (overE > 0) {
      const es = [...this.edges.values()].sort(
        (a, b) => a.weight - b.weight || a.lastAccess - b.lastAccess
      );
      for (let i = 0; i < overE && i < es.length; i++) {
        this.removeEdge(es[i].id);
      }
    }
  }
  private removeNode(id: string) {
    const n = this.nodes.get(id);
    if (!n) {
      return;
    }
    for (const e of [...this.edges.values()]) {
      if (e.from === id || e.to === id) {
        this.removeEdge(e.id);
      }
    }
    this.nodes.delete(id);
    if (n.canonicalKey) {
      this.byCanonical.delete(n.canonicalKey);
    }
  }
  private removeEdge(id: string) {
    const e = this.edges.get(id);
    if (!e) {
      return;
    }
    const a = this.nodes.get(e.from);
    if (a) {
      a.degree = Math.max(0, a.degree - 1);
    }
    const b = this.nodes.get(e.to);
    if (b) {
      b.degree = Math.max(0, b.degree - 1);
    }
    this.edges.delete(id);
  }

  stats() {
    return { nodes: this.nodes.size, edges: this.edges.size };
  }

  toSnapshot() {
    const nodes = [...this.nodes.values()].map((n) => ({
      ...n,
      embedding: serializeSparse(n.embedding),
    }));
    const edges = [...this.edges.values()];
    const embedder = this.embedder.exportState?.() ?? null;
    return { graph: { nodes, edges }, embedder, canonicalMeta: defaultCanonicalMeta };
  }
  fromSnapshot(s: { graph: { nodes: Array<{ id: string; embedding: unknown; [k: string]: unknown }>; edges: Edge[] }; embedder?: unknown }) {
    this.nodes.clear();
    this.edges.clear();
    this.byCanonical.clear();
    if (this.inverted) {
      this.inverted.clear();
    }
    for (const raw of s.graph.nodes) {
      const n = { ...(raw as Record<string, unknown>), embedding: deserializeSparse(raw.embedding) } as BaseNode;
      this.nodes.set(n.id, n);
      this.indexNode(n.id, n.text, n.canonicalKey);
    }
    for (const e of s.graph.edges) {
      this.edges.set(e.id, e);
    }
    this.embedder.importState?.(s.embedder);
  }
  async saveToFile(file: string, opts?: { onlyIfDirty?: boolean }) {
    if (opts?.onlyIfDirty && !this.dirty) {
      return;
    }
    await fs.writeFile(file, JSON.stringify(this.toSnapshot()), 'utf8');
    this.dirty = false;
  }
  static async loadFromFile(file: string, store: GraphStore) {
    const json = await fs.readFile(file, 'utf8');
    store.fromSnapshot(JSON.parse(json));
  }
}
