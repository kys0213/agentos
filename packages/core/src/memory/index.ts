export * from './graph-store';
export * from './memory-orchestrator';
export * from './types';
