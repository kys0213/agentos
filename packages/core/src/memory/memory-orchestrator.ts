import path from 'path';
import fs from 'fs/promises';
import {
  GraphStore,
  SleepCycleBaseOptions,
  SleepCycleReport,
  SleepCycleRunOptions,
  SleepCycleSummarizer,
} from './graph-store';
import { GraphConfig, BaseNode, Edge } from './types';
import { SimpleEmbedding } from './embedding/simple-embedding';

export type Scope = 'agent' | 'session';

export interface OrchestratorCfg {
  sessionGraph: GraphConfig;
  agentGraph: GraphConfig;
  promotion: { minRank: number; maxPromotions: number; minDegree?: number; carryWeights?: boolean };
  checkpoint: { outDir: string; topK: number; pretty?: boolean };
  searchBiasSessionFirst?: number;
  sleepCycle?: {
    summarizer: SleepCycleSummarizer;
    defaults?: Partial<SleepCycleBaseOptions>;
    runOnFinalize?: boolean;
    runOptions?: SleepCycleRunOptions;
    emit?: (scope: Scope, report: SleepCycleReport) => void;
  };
}

export class MemoryOrchestrator {
  private embedder = new SimpleEmbedding();
  private agentStore: GraphStore;
  private sessions = new Map<string, GraphStore>();

  constructor(
    private agentId: string,
    private cfg: OrchestratorCfg
  ) {
    this.agentStore = new GraphStore({ ...cfg.agentGraph }, this.embedder);
    this.configureSleepCycle(this.agentStore, 'agent');
  }

  getSessionStore(sessionId: string) {
    if (!this.sessions.has(sessionId)) {
      const store = new GraphStore({ ...this.cfg.sessionGraph }, this.embedder);
      this.configureSleepCycle(store, 'session');
      this.sessions.set(sessionId, store);
    }
    return this.sessions.get(sessionId)!;
  }
  getAgentStore() {
    return this.agentStore;
  }
  getActiveSessions() {
    return [...this.sessions.keys()];
  }

  upsertQuery(sessionId: string, text: string) {
    return this.getSessionStore(sessionId).upsertQuery(text);
  }
  recordFeedback(sessionId: string, qId: string, label: 'up' | 'down' | 'retry', note?: string) {
    return this.getSessionStore(sessionId).recordFeedback(qId, label, note);
  }

  search(sessionId: string, text: string, k = 8) {
    const s = this.getSessionStore(sessionId);
    const sTop = s.searchSimilarQueries(text, k);
    const aTop = this.agentStore.searchSimilarQueries(text, k);
    const bias = this.cfg.searchBiasSessionFirst ?? 0.05;
    const map = new Map<
      string,
      { id: string; score: number; from: 'session' | 'agent'; text?: string; canonicalKey?: string }
    >();
    for (const r of sTop) {
      map.set(r.canonicalKey ?? r.id, {
        id: r.id,
        score: r.score + bias,
        from: 'session',
        text: r.text,
        canonicalKey: r.canonicalKey,
      });
    }
    for (const r of aTop) {
      const key = r.canonicalKey ?? r.id;
      const cur = map.get(key);
      if (!cur) {
        map.set(key, {
          id: r.id,
          score: r.score,
          from: 'agent',
          text: r.text,
          canonicalKey: r.canonicalKey,
        });
      } else {
        cur.score = Math.max(cur.score, r.score + bias);
      }
    }
    return [...map.values()].sort((a, b) => b.score - a.score).slice(0, k);
  }

  private configureSleepCycle(store: GraphStore, scope: Scope) {
    if (!this.cfg.sleepCycle) {
      return;
    }
    store.configureSleepCycle({
      summarizer: this.cfg.sleepCycle.summarizer,
      defaults: this.cfg.sleepCycle.defaults,
      emit: this.cfg.sleepCycle.emit
        ? (report) => this.cfg.sleepCycle?.emit?.(scope, report)
        : undefined,
    });
  }

  private rank(n: Pick<BaseNode, 'lastAccess' | 'weights'>, halfLifeMin: number) {
    const now = Date.now();
    const ageMin = (now - n.lastAccess) / (1000 * 60);
    const recency = Math.exp(-Math.log(2) * (ageMin / halfLifeMin));
    const repeat = Math.tanh((n.weights?.repeat ?? 0) / 5);
    const fb = Math.tanh((n.weights?.feedback ?? 0) / 5);
    return 0.5 * recency + 0.3 * repeat + 0.2 * fb;
  }
  private pickTopQueries(
    nodes: BaseNode[],
    minRank: number,
    topK: number,
    minDegree: number,
    halfLifeMin: number
  ) {
    return nodes
      .filter((n) => n.type === 'query')
      .map((n) => ({ n, r: this.rank(n, halfLifeMin) }))
      .filter((x) => x.r >= minRank && (x.n.degree ?? 0) >= minDegree)
      .sort((a, b) => b.r - a.r)
      .slice(0, topK)
      .map((x) => x.n);
  }
  private cpFile(name: string) {
    return path.join(this.cfg.checkpoint.outDir, name);
  }

  async finalizeSession(
    sessionId: string,
    opts?: { promote?: boolean; checkpoint?: boolean; checkpointName?: string }
  ) {
    const promote = opts?.promote ?? true;
    const checkpoint = opts?.checkpoint ?? true;
    const s = this.sessions.get(sessionId);
    if (!s) {
      return;
    }

    if (this.cfg.sleepCycle && this.cfg.sleepCycle.runOnFinalize !== false) {
      const overrides: SleepCycleRunOptions = {
        ...(this.cfg.sleepCycle.runOptions ?? {}),
        trigger: 'session-finalize',
        force: true,
      };
      s.runSleepCycle(overrides);
    }

    if (promote) {
      const promoted = this.promoteHotspotsFromSession(sessionId, {
        carryWeights: this.cfg.promotion.carryWeights,
      });
      if (promoted > 0) {
        await fs.mkdir(this.cfg.checkpoint.outDir, { recursive: true });
        await this.agentStore.saveToFile(this.cpFile(`agent-${this.agentId}.json`), {
          onlyIfDirty: true,
        });
      }
    }

    if (checkpoint) {
      const snap = s.toSnapshot();
      const top = this.pickTopQueries(
        this.normalizeNodes(
          snap.graph.nodes as Array<Omit<BaseNode, 'embedding'> & { embedding: unknown }>
        ),
        this.cfg.promotion.minRank,
        this.cfg.checkpoint.topK,
        this.cfg.promotion.minDegree ?? 0,
        this.cfg.sessionGraph.halfLifeMin
      );
      const summary = {
        version: 1,
        agentId: this.agentId,
        sessionId,
        createdAt: new Date().toISOString(),
        topQueries: top.map((q) => ({
          id: q.id,
          text: q.text,
          lastAccess: q.lastAccess,
          weights: q.weights,
          degree: q.degree,
          canonicalKey: q.canonicalKey,
        })),
        edges: (snap.graph.edges as Edge[])
          .filter((e) => top.some((q) => q.id === e.from || q.id === e.to))
          .map((e) => ({ from: e.from, to: e.to, type: e.type, weight: e.weight })),
        embedder: snap.embedder,
        canonicalMeta: snap.canonicalMeta,
      };
      await fs.mkdir(this.cfg.checkpoint.outDir, { recursive: true });
      const name =
        opts?.checkpointName ?? `checkpoint-${this.agentId}-${sessionId}-${Date.now()}.json`;
      await fs.writeFile(
        this.cpFile(name),
        JSON.stringify(summary, null, this.cfg.checkpoint.pretty ? 2 : 0),
        'utf8'
      );
    }
    this.sessions.delete(sessionId);
  }

  promoteHotspotsFromSession(sessionId: string, opts?: { carryWeights?: boolean }) {
    const s = this.sessions.get(sessionId);
    if (!s) {
      return 0;
    }
    const snap = s.toSnapshot();
    const top = this.pickTopQueries(
      this.normalizeNodes(
        snap.graph.nodes as Array<Omit<BaseNode, 'embedding'> & { embedding: unknown }>
      ),
      this.cfg.promotion.minRank,
      this.cfg.promotion.maxPromotions,
      this.cfg.promotion.minDegree ?? 0,
      this.cfg.sessionGraph.halfLifeMin
    );
    let count = 0;
    for (const n of top) {
      if (n.type !== 'query' || !n.text) {
        continue;
      }
      const id = this.agentStore.upsertQuery(n.text);
      if (opts?.carryWeights) {
        const delta = (n.weights?.feedback ?? 0) + (n.weights?.repeat ?? 0) * 0.2;
        this.agentStore.adjustWeights(id, { feedbackDelta: delta });
      }
      count++;
    }
    return count;
  }

  private normalizeNodes(
    nodes: Array<Omit<BaseNode, 'embedding'> & { embedding: unknown }>
  ): BaseNode[] {
    return nodes.map((n) => {
      const e = (n as { embedding?: unknown }).embedding;
      let embedding: number[] | Map<number, number> | undefined;
      if (e == null) {
        embedding = undefined;
      } else if (Array.isArray(e)) {
        if (e.length === 0 || typeof e[0] === 'number') {
          embedding = e as number[];
        } else {
          try {
            embedding = new Map(e as [number, number][]);
          } catch {
            embedding = undefined;
          }
        }
      } else if (e instanceof Map) {
        embedding = e as Map<number, number>;
      } else {
        embedding = undefined;
      }
      return { ...n, embedding } as BaseNode;
    });
  }
}
