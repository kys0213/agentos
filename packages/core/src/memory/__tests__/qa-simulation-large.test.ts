import { MemoryOrchestrator } from '../../memory/memory-orchestrator';

const cfg = {
  sessionGraph: {
    maxNodes: 1000,
    maxEdges: 12000,
    halfLifeMin: 240,
    tauDup: 0.96,
    tauSim: 0.75,
    protectMinDegree: 3,
    enableInvertedIndex: true,
  },
  agentGraph: {
    maxNodes: 1000,
    maxEdges: 12000,
    halfLifeMin: 1440,
    tauDup: 0.96,
    tauSim: 0.78,
    protectMinDegree: 4,
    enableInvertedIndex: true,
  },
  promotion: { minRank: 0.55, maxPromotions: 40, minDegree: 1, carryWeights: true },
  checkpoint: { outDir: './.agentos/checkpoints', topK: 40, pretty: false },
  searchBiasSessionFirst: 0.05,
};

function buildDataset(): string[] {
  const products = ['제품A', '제품B', '제품C', '제품D'];
  const authoring = [
    (p: string) => `${p} 테스트 케이스 작성`,
    (p: string) => `${p} 테스트케이스 작성`,
    (p: string) => `${p} 테스트 시나리오 작성`,
  ];
  const regression = [
    (p: string) => `${p} 리그레이션 테스트 수행`,
    (p: string) => `${p} 회귀 테스트 실행`,
  ];
  const retrieval = [
    (p: string) => `${p} 작성된 테스트케이스 조회`,
    (p: string) => `${p} 테스트케이스 목록 조회`,
    (p: string) => `${p} 테스트 케이스 검색`,
  ];
  const evidence = [
    (p: string) => `${p} 테스트 증적 검증 요청`,
    (p: string) => `${p} 테스트 결과 증빙 확인`,
  ];
  const search = [
    (p: string) => `${p} 개편 관련 테스트케이스 있어?`,
    (p: string) => `${p} 리팩토링 회귀 테스트 계획`,
  ];

  const out: string[] = [];
  for (const p of products) {
    for (const fn of authoring) {
      out.push(fn(p));
    }
    for (const fn of regression) {
      out.push(fn(p));
    }
    for (const fn of retrieval) {
      out.push(fn(p));
    }
    for (const fn of evidence) {
      out.push(fn(p));
    }
    for (const fn of search) {
      out.push(fn(p));
    }
  }
  // paraphrases without product prefix
  out.push(
    '테스트 케이스 초안 만들기',
    '테스트케이스 템플릿 생성',
    '회귀 테스트 계획 수립',
    '증적 확인 요청',
    '테스트 증적 검토'
  );

  // repeat to reach ~100
  return Array(3)
    .fill(0)
    .flatMap(() => out);
}

describe('QA agent large simulation (~100 queries)', () => {
  test('graph forms clusters and promotions carry weights [snapshot]', async () => {
    // Fix time to ensure deterministic ranks/scores
    let tick = 1_700_000_000_000;
    const nowSpy = jest.spyOn(Date, 'now').mockImplementation(() => (tick += 1000));
    const o = new MemoryOrchestrator('qa-agent-large', cfg);
    const sid = 's-qa-large';
    const dataset = buildDataset().slice(0, 100);

    for (const q of dataset) {
      const id = o.upsertQuery(sid, q);
      if (q.includes('작성') || q.includes('생성')) {
        o.recordFeedback(sid, id, 'up');
      }
      if (q.includes('리팩토링')) {
        o.recordFeedback(sid, id, 'retry');
      }
      if (q.includes('증적')) {
        o.recordFeedback(sid, id, 'up');
      }
    }

    const sStore = o.getSessionStore(sid);
    const sStats = sStore.stats();
    expect(sStats.nodes).toBeGreaterThan(30);
    expect(sStats.edges).toBeGreaterThan(20);

    // Snapshot summary with structural metrics
    const snap = sStore.toSnapshot();
    type RawNode = {
      type?: string;
      id: string;
      text?: string;
      degree?: number;
      weights?: { feedback?: number; repeat?: number };
    };
    const edges = snap.graph.edges;
    const nodes = (snap.graph.nodes as RawNode[]).filter((n) => n.type === 'query');

    const typeCounts = edges.reduce(
      (acc: Record<string, number>, e) => {
        acc[e.type] = (acc[e.type] ?? 0) + 1;
        return acc;
      },
      {} as Record<string, number>
    );
    const degreeArr = nodes.map((n) => n.degree ?? 0);
    const fbArr = nodes.map((n) => Number(n.weights?.feedback ?? 0));
    const rpArr = nodes.map((n) => Number(n.weights?.repeat ?? 0));

    const quant = (arr: number[]) => {
      if (arr.length === 0) {
        return { min: 0, max: 0, mean: 0, median: 0, p90: 0 };
      }
      const sorted = [...arr].sort((a, b) => a - b);
      const sum = arr.reduce((a, b) => a + b, 0);
      const mean = sum / arr.length;
      const mid = Math.floor((sorted.length - 1) / 2);
      const median = sorted[mid];
      const p90 = sorted[Math.floor(0.9 * (sorted.length - 1))];
      const round = (x: number) => Number(x.toFixed(2));
      return {
        min: sorted[0],
        max: sorted[sorted.length - 1],
        mean: round(mean),
        median: round(median),
        p90: round(p90),
      };
    };

    // Connected components on similar_to as undirected
    const adj = new Map<string, Set<string>>();
    for (const e of edges) {
      if (e.type === 'similar_to') {
        if (!adj.has(e.from)) {
          adj.set(e.from, new Set());
        }
        if (!adj.has(e.to)) {
          adj.set(e.to, new Set());
        }
        adj.get(e.from)!.add(e.to);
        adj.get(e.to)!.add(e.from);
      }
    }
    const seen = new Set<string>();
    const compSizes: number[] = [];
    const nodeIds = nodes.map((n) => n.id);
    for (const id of nodeIds) {
      if (!seen.has(id)) {
        let size = 0;
        const q = [id];
        seen.add(id);
        while (q.length) {
          const cur = q.pop()!;
          size++;
          for (const nb of adj.get(cur)?.values() ?? []) {
            if (!seen.has(nb)) {
              seen.add(nb);
              q.push(nb as string);
            }
          }
        }
        compSizes.push(size);
      }
    }
    compSizes.sort((a, b) => b - a);

    const metrics = {
      counts: { nodes: sStats.nodes, edges: sStats.edges, queries: nodes.length },
      edgeTypes: typeCounts,
      similarRatio: Number(
        ((typeCounts['similar_to'] ?? 0) / Math.max(1, edges.length)).toFixed(3)
      ),
      degree: quant(degreeArr),
      feedback: quant(fbArr),
      repeat: quant(rpArr),
      components: { count: compSizes.length, top: compSizes.slice(0, 5) },
      topTexts: nodes
        .sort((a, b) => Number(b.weights?.feedback ?? 0) - Number(a.weights?.feedback ?? 0))
        .slice(0, 10)
        .map((n) => ({
          text: n.text,
          fb: Number((n.weights?.feedback ?? 0).toFixed(2)),
          degree: n.degree,
        })),
    };
    expect(metrics).toMatchSnapshot();

    await o.finalizeSession(sid, { promote: true, checkpoint: false });
    const aStore = o.getAgentStore();
    const aStats = aStore.stats();
    expect(aStats.nodes).toBeGreaterThan(0);

    const aSnap = aStore.toSnapshot();
    const aEdges = aSnap.graph.edges;
    const aNodes = (aSnap.graph.nodes as RawNode[]).filter((n) => n.type === 'query');
    const aTypeCounts = aEdges.reduce(
      (acc: Record<string, number>, e) => {
        acc[e.type] = (acc[e.type] ?? 0) + 1;
        return acc;
      },
      {} as Record<string, number>
    );
    const aMetrics = {
      counts: { nodes: aStats.nodes, edges: aStats.edges, queries: aNodes.length },
      edgeTypes: aTypeCounts,
      similarRatio: Number(
        ((aTypeCounts['similar_to'] ?? 0) / Math.max(1, aEdges.length)).toFixed(3)
      ),
      topTexts: aNodes
        .sort((a, b) => Number(b.weights?.feedback ?? 0) - Number(a.weights?.feedback ?? 0))
        .slice(0, 10)
        .map((n) => ({
          text: n.text,
          fb: Number((n.weights?.feedback ?? 0).toFixed(2)),
          degree: n.degree,
        })),
      search: aStore
        .searchSimilarQueries('제품C 리그레이션 테스트', 10)
        .map((r) => ({ text: r.text, score: Number(r.score.toFixed(2)) }))
        .sort((a, b) => (a.text || '').localeCompare(b.text || '')),
    };
    expect(aMetrics).toMatchSnapshot();
    nowSpy.mockRestore();
  });
});
