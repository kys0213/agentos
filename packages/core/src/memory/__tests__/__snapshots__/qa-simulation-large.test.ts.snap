// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QA agent large simulation (~100 queries) graph forms clusters and promotions carry weights [snapshot] 1`] = `
{
  "components": {
    "count": 25,
    "top": [
      4,
      4,
      4,
      4,
      4,
    ],
  },
  "counts": {
    "edges": 102,
    "nodes": 103,
    "queries": 53,
  },
  "degree": {
    "max": 5,
    "mean": 2.91,
    "median": 3,
    "min": 0,
    "p90": 5,
  },
  "edgeTypes": {
    "has_feedback": 50,
    "similar_to": 52,
  },
  "feedback": {
    "max": 2.9000000000000004,
    "mean": 1.33,
    "median": 0.9,
    "min": -0.10000000000000003,
    "p90": 2.9,
  },
  "repeat": {
    "max": 1,
    "mean": 0.89,
    "median": 1,
    "min": 0,
    "p90": 1,
  },
  "similarRatio": 0.51,
  "topTexts": [
    {
      "degree": 5,
      "fb": 2.9,
      "text": "제품A 테스트 시나리오 작성",
    },
    {
      "degree": 5,
      "fb": 2.9,
      "text": "제품A 작성된 테스트케이스 조회",
    },
    {
      "degree": 5,
      "fb": 2.9,
      "text": "제품A 테스트 증적 검증 요청",
    },
    {
      "degree": 5,
      "fb": 2.9,
      "text": "제품B 테스트 시나리오 작성",
    },
    {
      "degree": 5,
      "fb": 2.9,
      "text": "제품B 작성된 테스트케이스 조회",
    },
    {
      "degree": 5,
      "fb": 2.9,
      "text": "제품B 테스트 증적 검증 요청",
    },
    {
      "degree": 5,
      "fb": 2.9,
      "text": "제품C 테스트 시나리오 작성",
    },
    {
      "degree": 5,
      "fb": 2.9,
      "text": "제품C 작성된 테스트케이스 조회",
    },
    {
      "degree": 5,
      "fb": 2.9,
      "text": "제품C 테스트 증적 검증 요청",
    },
    {
      "degree": 5,
      "fb": 2.9,
      "text": "제품D 테스트 시나리오 작성",
    },
  ],
}
`;

exports[`QA agent large simulation (~100 queries) graph forms clusters and promotions carry weights [snapshot] 2`] = `
{
  "counts": {
    "edges": 17,
    "nodes": 40,
    "queries": 40,
  },
  "edgeTypes": {
    "similar_to": 17,
  },
  "search": [
    {
      "score": 0.62,
      "text": "제품A 리그레이션 테스트 수행",
    },
    {
      "score": 0.59,
      "text": "제품B 리그레이션 테스트 수행",
    },
    {
      "score": 0.78,
      "text": "제품C 리그레이션 테스트 수행",
    },
    {
      "score": 0.31,
      "text": "제품C 작성된 테스트케이스 조회",
    },
    {
      "score": 0.31,
      "text": "제품C 테스트 시나리오 작성",
    },
    {
      "score": 0.31,
      "text": "제품C 테스트 증적 검증 요청",
    },
    {
      "score": 0.29,
      "text": "제품C 테스트 케이스 검색",
    },
    {
      "score": 0.31,
      "text": "제품C 테스트 케이스 작성",
    },
    {
      "score": 0.31,
      "text": "제품C 테스트케이스 작성",
    },
    {
      "score": 0.59,
      "text": "제품D 리그레이션 테스트 수행",
    },
  ],
  "similarRatio": 1,
  "topTexts": [
    {
      "degree": 3,
      "fb": 4,
      "text": "제품B 작성된 테스트케이스 조회",
    },
    {
      "degree": 3,
      "fb": 4,
      "text": "제품C 작성된 테스트케이스 조회",
    },
    {
      "degree": 3,
      "fb": 4,
      "text": "제품A 작성된 테스트케이스 조회",
    },
    {
      "degree": 3,
      "fb": 4,
      "text": "제품D 작성된 테스트케이스 조회",
    },
    {
      "degree": 0,
      "fb": 3.1,
      "text": "제품C 테스트 증적 검증 요청",
    },
    {
      "degree": 0,
      "fb": 3.1,
      "text": "제품C 테스트 시나리오 작성",
    },
    {
      "degree": 0,
      "fb": 3.1,
      "text": "제품B 테스트 증적 검증 요청",
    },
    {
      "degree": 0,
      "fb": 3.1,
      "text": "제품B 테스트 시나리오 작성",
    },
    {
      "degree": 0,
      "fb": 3.1,
      "text": "제품A 테스트 증적 검증 요청",
    },
    {
      "degree": 0,
      "fb": 3.1,
      "text": "제품A 테스트 시나리오 작성",
    },
  ],
}
`;
